designer/src/test/java/org/restcomm/connect/rvd/restcomm/RestcommClientTest.java

```diff
@@ -46,10 +60,52 @@ public static void init() {
     }
 
     @Ignore
-    @Test(expected=RestcommClient.RestcommClientInitializationException.class)
+    @Test(expected = RestcommClient.RestcommClientInitializationException.class)
     public void exceptionThrownWhenNoCredentialsCanBeDetermined() throws RestcommClient.RestcommClientInitializationException, URISyntaxException {
         CustomHttpClientBuilder httpClientBuilder = new CustomHttpClientBuilder(configuration);
-        RestcommClient client = new RestcommClient(null, null,httpClientBuilder);
+        RestcommClient client = new RestcommClient(null, null, httpClientBuilder.buildHttpClient());
+    }
+
+    @Test(expected = RestcommClient.RestcommClientException.class)
+    public void noMoreConnectionsInPool() throws Exception {
+        int connsPerRoute= 5;
+        int timeout  = 1000;
+        final String path = "/noMoreConnections";
+        List<RvdMaxPerHost> routes = new ArrayList();
+        routes.add(new RvdMaxPerHost("http://127.0.0.1:8099", connsPerRoute));
+        RvdConfiguration configuration = Mockito.mock(RvdConfiguration.class);
+        when(configuration.getDefaultHttpMaxConns()).thenReturn(connsPerRoute * 2);
+        when(configuration.getDefaultHttpMaxConnsPerRoute()).thenReturn(1);
+        when(configuration.getDefaultHttpTimeout()).thenReturn(timeout);
+        when(configuration.getDefaultHttpTTL()).thenReturn(3000);
+        when(configuration.getDefaultHttpMaxPerRoute()).thenReturn(routes);
+        CustomHttpClientBuilder httpClientBuilder = new CustomHttpClientBuilder(configuration);
+        stubFor(get(urlMatching(path)).willReturn(aResponse()
+                .withFixedDelay(timeout * 10)//delay will cause tiemout to happen on second request
+                .withStatus(200)
+                .withHeader("Content-Type", "application/json")
+                .withBody("{\"sid\":\"" + 1234 + "\",\"email_address\":\"" + "jander@clander.com" + "\",\"status\":\"active\",\"role\":\"administrator\"}")));
+        URI create = URI.create("http://127.0.0.1:8099");
+        final RestcommClient client = new RestcommClient(create, "authHeader", httpClientBuilder.buildHttpClient());
+        final Gson gson = new Gson();
+
+        for (int i = 0; i < connsPerRoute; i++) {
+            Thread newThread = new Thread() {
+                public void run() {
+                    try {
+                        RestcommAccountInfo done = client.get(path).done(gson, RestcommAccountInfo.class);
+                        Assert.fail("these request are expected to fail because of timeout.");
+                    } catch (AccessApiException ex) {
+                    }
+                }
+            };
+            newThread.start();
+        }
+        Thread.sleep(200);
+        client.get("/noMoreConnections").done(gson, RestcommAccountInfo.class);
```

Comment:
@jaimecasero , please tell me if i followed the rational of the test correctly:

You simulate a client with a pool of one (5) connections per route (at least for route http://127.0.0.1:8099). This means while the first five request are pending (server has not responded) any subsequent request will fail. Correct ? So, you create a set of threads that occupy first 5 available client connection slots silently. At the end (outside of the for()) you make yet another request and expect this to fail bacause there are no more client connection slots. Is this correct ?

I think i miss something. If  you configure wiremock to delay for timeout*10 i.e. 10 sec, the final request will fail (and the test will pass) even without any threads occupying request slots. And the reason will _not be_ that there are no more request slots/connections in the pool, but because wiremock has a greater timeout than client (10000 > 1000). I confirmed the hypothesis by indeed commenting out the connections within the threads. The test still passed (read timed out was thrown when removed the 'expected' stuff).

Are you sure we're indeed testing the availability of connections ?

Also, just to help get in the loop, what would be the proper behavior in case there connection limit is reached for a route ? Throw an exception or block the thread ?